public/static/app/perfdiscovery/PerfDiscoveryService.js

```diff
@@ -0,0 +1,207 @@
+mciModule.factory('PerfDiscoveryService', function($q, ApiV1, ApiTaskdata) {
+  var _respData = function(resp) {
+    return resp.data
+  }
+
+  // TODO Should be part of `VersionResource`
+  // :parm version: version object
+  // :returns: Array of {taskId, taskName, buildName}
+  function extractTasks(version) {
+    return _.reduce(version.builds, function(items, build) {
+      // Going over dict of tasks ({taskName: data})
+      return items.concat(_.map(build.tasks, function(taskData, taskName) {
+        return {
+          taskId: taskData.task_id,
+          taskName: taskName,
+          buildName: build.name,
+        }
+      }))
+    }, [])
+  }
+
+  // Makes series of HTTP calls and loads curent, baseline and history data
+  // :rtype: $q.promise
+  function queryData(tasks, version, baseline) {
+    var current = $q.all(_.map(tasks, function(task) {
+      return ApiTaskdata.getTaskById(task.taskId, 'perf')
+        .then(_respData, function() { return null })
+        .then(function(data) {
```

Comment:
Are you sure this is the correct syntax for callbacks in Angular 1.6? We're doing .then(successDelegate, failureDelegate) elsewhere

Reply:
I am sure. failure callback is optional.